kitsune/kpi/cron.py

```diff
@@ -496,3 +496,92 @@ def survey_recent_askers():
+    for _ in range(12):
+        previous_week = boundaries[-1] - timedelta(weeks=1)
+        boundaries.append(previous_week)
+    boundaries.reverse()
+    ranges = zip(boundaries[:-1], boundaries[1:])
+
+    reports = [
+        (CONTRIBUTOR_COHORT_CODE, [
+            (Revision.objects.all(), ('creator', 'reviewer',)),
+            (Answer.objects.not_by_asker(), ('creator',)),
+            (Reply.objects.all(), ('user',))]),
+        (KB_CONTRIBUTOR_COHORT_CODE, [
+            (Revision.objects.filter(document__locale='en-US'), ('creator', 'reviewer',))]),
+        (KB_L10N_CONTRIBUTOR_COHORT_CODE, [
+            (Revision.objects.exclude(document__locale='en-US'), ('creator', 'reviewer',))]),
+        (SUPPORT_FORUM_HELPER_COHORT_CODE, [
+            (Answer.objects.not_by_asker(), ('creator',))]),
+        (AOA_CONTRIBUTOR_COHORT_CODE, [
+            (Reply.objects.all(), ('user',))])
+    ]
+
+    for kind, querysets in reports:
+        cohort_kind = CohortKind.objects.get_or_create(code=kind)[0]
+
+        for i, cohort_range in enumerate(ranges):
+            cohort_users = _get_cohort(querysets, cohort_range)
+
+            # Sometimes None will be added to the cohort_users list, so remove it
+            if None in cohort_users:
+                cohort_users.remove(None)
+
+            cohort = Cohort.objects.get_or_create(kind=cohort_kind, start=cohort_range[0],
+                                                  end=cohort_range[1])[0]
+            cohort.size = len(cohort_users)
+            cohort.save()
+
+            for retention_range in ranges[i + 1:]:
+                retained_user_count = _count_contributors_in_range(
+                    querysets, cohort_users, retention_range)
+                retention_metric = RetentionMetric.objects.get_or_create(
+                    cohort=cohort, start=retention_range[0], end=retention_range[1])[0]
+                retention_metric.size = retained_user_count
+                retention_metric.save()
+
+
+def _count_contributors_in_range(querysets, users, date_range):
+    """Of the group ``users``, count how many made a contribution in ``date_range``."""
+    start, end = date_range
+    retained_users = set()
+
+    for queryset, fields in querysets:
+        for field in fields:
+            filters = {'%s__in' % field: users, 'created__gte': start, 'created__lt': end}
+            retained_users = retained_users.union(
+                set(getattr(o, field) for o in queryset.filter(**filters)))
+
+    return len(retained_users)
+
+
+def _get_cohort(querysets, date_range):
+    start, end = date_range
+    cohort = []
+
+    for queryset, fields in querysets:
+        contributions_in_range = queryset.filter(created__gte=start, created__lt=end)
+        potential_users = set()
+
+        for field in fields:
+            potential_users = potential_users.union(
+                set(getattr(cont, field) for cont in contributions_in_range))
```

Comment:
If you wanted, this could be `potential_users |= set(getattr(cont, field) for cont in contributions_in_range)`. The `|` operator is equivalent to `.union()`. `&` is `.intersection()`.